usaon_vta_survey/routes/response/relationships/observing_system_data_product.py

```diff
@@ -0,0 +1,222 @@
+from flask import Request, redirect, render_template, request, url_for
+from flask_wtf import FlaskForm
+from wtforms import FormField
+
+from usaon_vta_survey import app, db
+from usaon_vta_survey.forms import FORMS_BY_MODEL
+from usaon_vta_survey.models.tables import (
+    ResponseDataProduct,
+    ResponseObservingSystem,
+    ResponseObservingSystemDataProduct,
+    Survey,
+)
+from usaon_vta_survey.util.authorization import limit_response_editors
+
+
+def _update_super_form(
+    super_form: type[FlaskForm],
+    /,
+    *,
+    data_product_id: int | None,
+    observing_system_id: int | None,
+) -> None:
+    """Populate the form of forms with sub-forms depending on provided IDs.
+
+    When an ID for an object is not provided, we need to gather information from the
+    user to create that object.
+
+    TODO: Better function name.
+    """
+    if observing_system_id is None:
+        super_form.observing_system = FormField(FORMS_BY_MODEL[ResponseObservingSystem])
+
+    if data_product_id is None:
+        super_form.data_product = FormField(FORMS_BY_MODEL[ResponseDataProduct])
+
+
+def _update_relationship(
+    relationship: ResponseObservingSystemDataProduct,
+    *,
+    observing_system_id: int | None,
+    data_product_id: int | None,
+) -> None:
+    """Populate the relationship with any known identifiers.
+
+    TODO: Better function name.
+    """
+    if observing_system_id:
+        relationship.response_observing_system_id = observing_system_id
+
+    if data_product_id:
+        relationship.response_data_product_id = data_product_id
+
+
+# may not need to be internal
+def _response_data_product(
+    *,
+    data_product_id: int | None,
+    response_id: int,
+) -> ResponseDataProduct:
+    """Return a data product db object (or 404), and do some mutations.
+
+    TODO: Extract mutations to another function responsible for that.
+    """
+    if data_product_id is not None:
+        response_data_product = db.get_or_404(ResponseDataProduct, data_product_id)
+    else:
+        response_data_product = ResponseDataProduct(response_id=response_id)
+
+    return response_data_product
+
+
+def _response_observing_system(
+    *,
+    observing_system_id: int | None,
+    response_id: int,
+) -> ResponseObservingSystem:
+    """Return an observing system db object (or 404), and do some mutations."""
```

Comment:
Same as the function above. Maybe we could genericize things a bit since so many things depend on what arguments are coming in from the route.

Maybe something like this pseudocode:

```python
# TODO: Use TypeVar to ensure type of `model` is same as return type; replace `Any`s
def _get_or_404_or_init(model: Type[Model], identifiers: list[Any], init_params: dict[str, Any]) -> Model:
    """Query for record(s) matching `identifiers`, if provided.

    Otherwise, insert a new record with `init_params`. This pattern is useful for routes which deal with database relationships.

    E.g. a route which receives `response_id` (required) and `observing_system_id` (optional) would want to lookup or create the observing system based on whether it's passed. If the observing system needs to be created, we want to link it to the `response_id` that's required by the route.
    """
    if identifiers:
        record = db.get_or_404(model, tuple(identifiers))
    else:
        record = model(*init_params)
        # TODO: Should we actually add the record to the DB at this point?
        # EDIT: I think no; at the point we're calling these functions we don't know whether the request is a POST or not, so we don't know whether the DB should be modified. I don't like that this means that the function would return either a real DB object or a transient object, but we can keep thinking about how to do better :)
        # db.add(record)
        # db.commit()
    return record
```

Maybe after we replace multiple functions with this pattern, another repeated pattern will jump out that we can leverage. What do you think of this?

Again maybe this idea is more for the next PR than this one :)